### PR TITLE
Authentication Verification Refactor

### PR DESCRIPTION
**Pull request by Firstyear**
Supercedes #49 - this implements the start of the design as discussed in https://github.com/kanidm/webauthn-rs/blob/master/designs/authentication-use-cases.md

This allows:

* Multiple credentials, with a derive UV policy and checked for consistency.
* A single credential, which can have a policy over-ride applied.

- [ x ] cargo fmt has been run
- [ x ] cargo test has been run and passes
- [ - ] documentation has been updated with relevant examples (if relevant)

### CODE REVIEW

File: src/core.rs
Comment (agrinman) [inline src/core.rs:117]: any reason we  don't  just pass the policy directly here?

File: src/core.rs
Comment (agrinman) [inline src/core.rs:600]: I think this might cause some false positive errors. perhaps we want to control this with a config?
i.e: `if cred.verified && self.config.get_require_uv_consistency()`?

File: src/core.rs
Comment (agrinman) [inline src/core.rs:678]: why would we allow overriding the policy? this seems like an easy footgun

**Review by agrinman — COMMENTED** [review]
I think it's close, just a couple of comments above for discussion.

One other suggestion: let's add tests for our UV test cases because it's  very not obvious :) I.e. tests that check all the combinations for setting `credential.verified` and `policy` and ensuring that `register_internal` and `verify_internal` work as intended.

File: src/core.rs
Comment (Firstyear) [reply src/core.rs:117]: This is the "basic" register function, so I thought to keep it as UV=true/false for discouraged/required to make it simpler to hold. The policy can be directly used in the options version.

File: src/core.rs
Comment (Firstyear) [reply src/core.rs:678]: I seem to remember you requested the ability to override the UV policy for a singular credential for some use cases :) That's why I added it because I think for a single credential, it's still a footgun, but it's not an automatic foot chainsaw like policy on a set of credentials is.

File: src/core.rs
Comment (Firstyear) [reply src/core.rs:600]: Yep, this is a good point. I'll add that switch now.

**Comment by Firstyear** [comment]
Just need to add the tests later of the various combinations. :)

**Comment by Firstyear** [comment]
Updated with some tests of the UV consistency options.

**Review by ericmarkmartin — APPROVED** [review]

**Comment by Firstyear** [comment]
I'll let @agrinman have another look before I merge :)

File: src/core.rs
Comment (agrinman) [reply src/core.rs:685]: Perhaps I misread, but I thought I saw that in `verify_credential_internal` it would error if mismatched.
I think that the right way to "override" it is to just change it on the credential itself.

**Review by agrinman — APPROVED** [review]

File: src/core.rs
Comment (agrinman) [reply src/core.rs:685]: nvm ignore this!

Comment (Firstyear) [reply src/core.rs:685]: Yeah, when you set the override we don't do the policy consistency check. The only thing that *could* occur, is that the credential will verify under discouraged which we have a seperate flag for checking :)

**Pull request merged**
